### ISSUE TITLE
lnrpc: overlapping REST endpoints

### ISSUE THREAD

**Issue by ndeet** [lightningnetwork/lnd#544]
The currently generated `rpc.swagger.json` has overlapping endpoints which results in an error on importing the file. See:
https://app.swaggerhub.com/apis/lnd-rest/rpc-proto/master

The problem is that in the `rpc.proto` we have different classes defining the same REST endpoints but with different parameters:
```
 /** lncli: `listinvoices`
    ListInvoices returns a list of all the invoices currently stored within the
    database. Any active debug invoices are ignored.
    */
    rpc ListInvoices (ListInvoiceRequest) returns (ListInvoiceResponse) {
        option (google.api.http) = {
            get: "/v1/invoices/{pending_only}"
        };
    }

    /** lncli: `lookupinvoice`
    LookupInvoice attemps to look up an invoice according to its payment hash.
    The passed payment hash *must* be exactly 32 bytes, if not, an error is
    returned.
    */
    rpc LookupInvoice (PaymentHash) returns (Invoice) {
        option (google.api.http) = {
            get: "/v1/invoices/{r_hash_str}"
        };
    }
``` 

This ends up being the same endpoints but with different parameters which can't be distinguished:
- `get: "/v1/invoices/{pending_only}"`
- `get: "/v1/invoices/{r_hash_str}"`

For `ListChannels` there is a similar situation and it was solved there by keeping two separate endpoints (notice the //TODO which would also cause same error for 'ListChannels` REST endpoints):
```
    // TODO(roasbeef): merge with below with bool?
    /** lncli: `pendingchannels`
    PendingChannels returns a list of all the channels that are currently
    considered "pending". A channel is pending if it has finished the funding
    workflow and is waiting for confirmations for the funding txn, or is in the
    process of closure, either initiated cooperatively or non-cooperatively.
    */
    rpc PendingChannels (PendingChannelRequest) returns (PendingChannelResponse) {
        option (google.api.http) = {
           get: "/v1/channels/pending"
        };
    }

    /** lncli: `listchannels`
    ListChannels returns a description of all the open channels that this node
    is a participant in.
    */
    rpc ListChannels (ListChannelsRequest) returns (ListChannelsResponse) {
        option (google.api.http) = {
            get: "/v1/channels"
        };
    }
```

Solution 1:
The most consistent solution for now would be to add an additional pending endpoint (`get: "/v1/invoices/{pending_only}"`) to `ListInvoices` like it is already for `ListChannels`, imo. 

Solution 2:
Or alternatively to to change REST endpoints to support query parameters (`?pending=true`). But not sure if this is wanted or if the REST API should stay path based.

If maintainers can give some directions I will prepare a PR.

**Comment by Richard87** [comment]
Hi, I like Solution 2, as this would help with closeChannel to, to add a optional `?force=true` parameter...

There us also the problem with Subscribe Invoices (https://lightningcommunity.slack.com/archives/C6BDA6DGE/p1515062958000279)

**Comment by ndeet** [comment]
@Richard87 
Solution 2: 
yes seems also cleaner to me but will wait what maintainers prefer, if it should be path based we will have similar endpoints for each option like mentioned above and it will work if it there are rarely such flags needed. Otherwise it will get messy if we force everything into paths. 

> There us also the problem with Subscribe Invoices 
> (https://lightningcommunity.slack.com/archives/C6BDA6DGE/p1515062958000279)

I think this is caused also by `/v1/invoices/{pending_only}` as it takes up the third path segment, so the REST proxy does not understand it. It assumes always the `/v1/invoices/{pending_only}` endpoint no matter what is the third argument (because of variable segment):
`/v1/invoices/subscribe` => `/v1/invoices/{pending_only}` 
`/v1/invoices/pending` => `/v1/invoices/{pending_only}`
`/v1/invoices/something` => `/v1/invoices/{pending_only}`

**Comment by ndeet** [comment]
Seems that PR #596 is addressing solution 2) mentioned above for the invoices endpoint. 

For consistency we then should also do that for the mentioned `/v1/channels/pending` endpoint.

**Comment by Roasbeef** [comment]
Fixed by #596.
